cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java

```diff
@@ -129,32 +119,112 @@ public void metadata(FullHttpRequest request, HttpResponder responder) throws Ex
     Closeables.closeQuietly(chunkResponder);
   }
 
-  private String getTopic(String topicConfig) {
-    int idx = topicConfig.lastIndexOf(':');
-    return idx < 0 ? cConf.get(topicConfig) : cConf.get(topicConfig.substring(0, idx)) + topicConfig.substring(idx + 1);
-  }
-
+  /**
+   * shuts down remote runtime
+   */
   @POST
   @Path("/shutdown")
   public void shutdown(FullHttpRequest request, HttpResponder responder) throws Exception {
     responder.sendString(HttpResponseStatus.OK, "Triggering shutdown down Runtime Http Server.");
     shutdownRunnable.run();
   }
 
-  private void fetchAndWriteMessages(JsonWriter jsonWriter, ByteBuf buffer,
-                                     ChunkResponder chunkResponder, String topic, int limit,
+  /**
+   * Decode consume request from avro binary format
+   */
+  private Map<Utf8, GenericRecord> decodeConsumeRequest(FullHttpRequest request) throws IOException {
+    Decoder decoder = DecoderFactory.get().directBinaryDecoder(new ByteBufInputStream(request.content()), null);
+    DatumReader<Map<Utf8, GenericRecord>> datumReader = new GenericDatumReader<>(
+      MonitorSchemas.V1.MonitorConsumeRequest.SCHEMA);
+
+    // Avro converts java String to UTF8 while deserializing strings.
+    return datumReader.read(null, decoder);
+  }
+
+  /**
+   * Encode response in avro binary format
+   */
+  private void encodeAndSendResponse(Map<Utf8, GenericRecord> consumeRequests, ChunkResponder chunkResponder,
+                                     ByteBuf buffer) throws IOException {
+    Encoder encoder = EncoderFactory.get().directBinaryEncoder(new ByteBufOutputStream(buffer), null);
+    encoder.writeMapStart();
+    encoder.setItemCount(consumeRequests.size());
+
+    for (Map.Entry<Utf8, GenericRecord> entry : consumeRequests.entrySet()) {
+      String topicConfig = String.valueOf(entry.getKey());
+      encoder.startItem();
+      encoder.writeString(topicConfig);
+      encoder.writeArrayStart();
+
+      if (!requestKeyToLocalTopic.containsKey(topicConfig)) {
+        requestKeyToLocalTopic.put(topicConfig, getTopic(topicConfig));
+      }
+
+      String topic = requestKeyToLocalTopic.get(topicConfig);
+
+      try {
+        int limit = (int) entry.getValue().get("limit");
+        Optional<String> messageId = Optional.ofNullable(String.valueOf(entry.getValue().get("messageId")))
+          .filter(s -> !s.equals("null"));
```

Comment:
Shouldn't match the `null` string. If messageId is optional in the request, make the schema field optional (by using Union type).